src/backends/policy_engine/souffle/datalog_lowering_visitor.cc

```diff
@@ -0,0 +1,103 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#include "src/backends/policy_engine/souffle/datalog_lowering_visitor.h"
+
+#include "src/common/logging/logging.h"
+#include "src/ir/attributes/attribute.h"
+#include "src/ir/attributes/int_attribute.h"
+#include "src/ir/attributes/string_attribute.h"
+#include "src/ir/datalog/operation.h"
+#include "src/ir/module.h"
+#include "src/ir/operator.h"
+
+namespace raksha::backends::policy_engine::souffle {
+
+using DatalogOperation = ir::datalog::Operation;
+using DatalogSymbol = ir::datalog::Symbol;
+using DatalogNumber = ir::datalog::Number;
+using DatalogAttributeList = ir::datalog::AttributeList;
+using DatalogOperandList = ir::datalog::OperandList;
+using DatalogIsOperationFact = ir::datalog::IsOperationFact;
+using DatalogAttribute = ir::datalog::Attribute;
+using DatalogAttributePayload = ir::datalog::AttributePayload;
+using DatalogStringAttributePayload = ir::datalog::StringAttributePayload;
+using DatalogNumberAttributePayload = ir::datalog::NumberAttributePayload;
+using IrOperation = ir::Operation;
+
+static DatalogAttributePayload GetPayloadForAttribute(ir::Attribute attr,
+                                                      ir::SsaNames &ssa_names) {
+  if (auto int_attr = attr.GetIf<ir::Int64Attribute>()) {
+    return DatalogNumberAttributePayload(DatalogNumber(int_attr->value()));
+  } else if (auto string_attr = attr.GetIf<ir::StringAttribute>()) {
+    return DatalogStringAttributePayload(DatalogSymbol(string_attr->value()));
+  } else {
+    LOG(FATAL) << "Unknown attribute kind.";
+  }
+  // Unreachable, just to placate compiler.
+  return DatalogStringAttributePayload(DatalogSymbol(""));
+}
+
+static DatalogAttribute TranslateIrNodeToDatalog(
+    std::pair<std::string, ir::Attribute> name_attr_pair,
+    ir::SsaNames &ssa_names) {
+  return DatalogAttribute(
+      DatalogSymbol(std::move(name_attr_pair.first)),
+      GetPayloadForAttribute(std::move(name_attr_pair.second), ssa_names));
+}
+
+static DatalogSymbol TranslateIrNodeToDatalog(ir::Value value,
+                                              ir::SsaNames &ssa_names) {
+  return DatalogSymbol(value.ToString(ssa_names));
+}
+
+// Turn a range of IR nodes in a C++ container into a Datalog-style linked
+// list.
+template <class ListT, class IrNodeIter>
+static ListT RangeToDatalogList(IrNodeIter ir_node_list_begin,
+                                IrNodeIter ir_node_list_end,
+                                ir::SsaNames &ssa_names) {
+  if (ir_node_list_begin == ir_node_list_end) return ListT();
+  auto head_in_dl = TranslateIrNodeToDatalog(*ir_node_list_begin, ssa_names);
+  return ListT(std::move(head_in_dl),
+               RangeToDatalogList<ListT>(++ir_node_list_begin, ir_node_list_end,
+                                         ssa_names));
```

Comment:
Use a loop instead of recursion?

Reply:
I'm using the recursion because the Datalog `ListT` will be constructed by `cons`ing a new head over and over. This is easy to expression using recursion, but awkward to express with a loop.

Reply:
Sounds good. I will defer it to you. BTW, I think it is ok to use std::accumulate as the side-effect restriction is only until C++11. (I think we discussed this on another review thread.)

Reply:
You are correct, I misread the description in the docs earlier, and `std::accumulate` can handle side effects as of C++11. Unfortunately, `std::accumulate` copies its accumulated value until C++20, so we can't use it here.